### DISCUSSION TITLE
New Kernel Won't Load

### DISCUSSION THREAD

**Discussion by gaia** [microsoft/WSL#8989] [General] [answered]
### Version

10.0.22000.1042

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

Kernel version: 5.10.102.1

### Distro Version

Ubuntu 22.04

### Images

![image](https://user-images.githubusercontent.com/87547/194670439-1eabaef4-1765-4ba4-afc9-b4b8e6bbbad5.png)

ok, let's restart:

![image](https://user-images.githubusercontent.com/87547/194670669-4e221ffd-8540-4bc1-a23d-751cb78ccabe.png)

### Repro Steps

```$ uname -r
5.10.102.1-microsoft-standard-WSL2```

`$ sudo needrestart`

same old kernel. Let's try from within:

```$ uname -r
5.10.102.1-microsoft-standard-WSL2```

Also tried restarting the LXSS service and rebooting windows.


### Expected Behavior

at least one of those restart attempts to let WSL use the new kernel

### Actual Behavior

pre-update kernel remains loaded

### Diagnostic Logs

**Comment by zcobol** [comment]
@gaia how did you install the new kernel, deb package or custom build? `5.15.0-48-generic` doesn't look like a WSL released kernel. The latest kernel version is `5.15.68.1-microsoft-standard-WSL2` in the `pre-release` version. Notice the branding string `microsoft-standard-WSL2` which is added to kernels released by WSL team.

It is not recommended to install kernels in a WSL instance. It's just a container.

**Comment by NotTheDr01ds** [comment]
Also @gaia, there are some formatting issues with your post (images inside code-block, etc.).  Would recommend fixing these to make it easier to read.

But definitely agree with @zcobol -- This appears to be working as intended.

You cannot update the kernel from *within* a distribution; it has to be done for all of WSL2.  Please see [the config doc](https://learn.microsoft.com/en-us/windows/wsl/wsl-config#configuration-setting-for-wslconfig) for information on how to use a different kernel with WSL2, but also realize that you'll need to make sure that the kernel that you are running has all of the necessary features included (Hyper-V, 9P support, and more).  The best way to ensure this is by starting from the [WSL2 kernel sources](https://github.com/microsoft/WSL2-Linux-Kernel).

**Comment by gaia** [comment]
> @gaia how did you install the new kernel, deb package or custom build? `5.15.0-48-generic` doesn't look like a WSL released kernel. The latest kernel version is `5.15.68.1-microsoft-standard-WSL2` in the `pre-release` version. Notice the branding string `microsoft-standard-WSL2` which is added to kernels released by WSL team.
> 
> It is not recommended to install kernels in a WSL instance. It's just a container.

I believe it came with `apt --with-new-pkgs upgrade` from within the instance. You are right, I was so focused on the .15 I didn't notice it wasn't an MS issued kernel. I will remove it.

A question remains: I've always used `--with-new-pkgs upgrade` with WSL, but this is the first time this happens. Did something change with WSL, Ubuntu, or maybe the way Ubuntu was installed (by me)?

![image](https://user-images.githubusercontent.com/87547/194724411-36412ca8-92a1-4e46-8f81-2fd0ac679b74.png)

**Comment by gaia** [comment]
The other Q is, how do I make use of https://github.com/microsoft/WSL2-Linux-Kernel/releases/tag/rolling-lts%2Fwsl%2F5.15.68.1? Do I need to install Ubuntu again?

**Comment by NotTheDr01ds** [comment, marked as answer]
> I've always used `--with-new-pkgs` upgrade with WSL, but this is the first time this happens. Did something change with WSL, Ubuntu, or maybe the way Ubuntu was installed (by me)?

I can't claim to be an Apt expert, but I'm guessing that some *other* package that you installed may have specified the kernel as a dependency?  I just did a `sudo apt --with-new-pkgs upgrade` on my (fairly) freshly installed 22.04 instance and didn't get a kernel with it.

In general, I'd recommend just a `sudo apt upgrade` unless you see a held-back package that you know you need and should install.

> The other Q is, how do I make use of [[the 5.15.68.1 kernel](https://github.com/microsoft/WSL2-Linux-Kernel/releases/tag/rolling-lts%2Fwsl%2F5.15.68.1)]? Do I need to install Ubuntu again?

That one's pretty easy.  Since you are running Windows 11, you can install WSL [0.68.4](https://github.com/microsoft/WSL/releases/tag/0.68.4), the latest pre-release that includes that kernel.

To do so:

* On the releases page linked above, expand "Assets" and download the `.msixbundle`

* Reboot to clear any in-use WSL files (may not be necessary, and there are other ways to handle it, but this is simply the easiest to explain :-).

* Start an Administrative PowerShell

* Navigate to the download folder in PowerShell

* Run `Add-AppxPackage Microsoft.WSL_0.68.4.0_x64_ARM64.msixbundle`

* `wsl --version` should show the latest kernel is installed.

**Comment by gaia** [comment]
Thanks, @NotTheDr01ds.

For others that find this, note that (you might want to update your post)

`Add-AppxPackage Microsoft.WSL_0.66.2.0_x64_ARM64.msixbundle`

`Add`, not `App`. Followed by `wsl --status`, not `wsl --version` 

Lastly, just out of curiosity, the output of --status changed considerably after installing the kernel

![image](https://user-images.githubusercontent.com/87547/194911659-acdfdffa-5237-442f-aeb3-865a0b4a475c.png)

**Comment by NotTheDr01ds** [comment]
@gaia 

> `Add`, not `App`

Good catch - Typo on my part.

> Followed by `wsl --status`, not `wsl --version`

Not a typo -- It really should be `--version` (the command changes again in 0.68.4), but from the screenshot it looks like you downloaded and installed 0.66.2 instead of 0.68.4.  They recently (0.67.6) tweaked the `--status` and `--version` commands to remove redundant information.

After installing 0.68.4, the results should be something like:

```
> wsl --status
Default Distribution: ntd_Tumbleweed
Default Version: 2
> wsl --version
WSL version: 0.68.4.0
Kernel version: 5.15.68.1
WSLg version: 1.0.44
MSRDC version: 1.2.3401
Direct3D version: 1.606.4
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.22621.608
```
